src/os/bluestore/BlueStore.cc

```diff
@@ -6361,6 +6370,12 @@ void BlueStore::_txc_state_proc(TransContext *txc)
 	return;
       }
       txc->state = TransContext::STATE_FINISHING;
+      if (txc->kv_submitted_sync && !txc->wal_txn) {
+	std::lock_guard<std::mutex> l(kv_lock);
+	kv_queue.push_back(txc);
+	kv_cond.notify_one();
+	return;
+      }
```

Comment:
@liewegas 
To call _txc_finish and _osr_reap_done by any thread other than ShardedOpWq, I added !txc->wal_txn condition.
### wal_txn case: 
Context1(ShardedOpWQ) -> **_txc_finish_io** -> _txc_finish_kv  -> _wal_apply -> _txc_aio_submit
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pg lock&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**
Context2(KernelDevice.aio_thread) -> _wal_finish
Context3(kv_sync_thread) -> _kv_sync_thread -> **_txc_finish** calling **_osr_reap_done**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**
There's no problem...
<br/>

### w/o !wal_txn condition in _txc_state_proc on STATE_KV_DONE case:
Context1(ShardedOpWQ) -> **_txc_finish_io** -> _txc_finish_kv  -> **_txc_finish** calling **_osr_reap_done**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pg lock&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**
There is a problem. ShardedOpWQ tries to take osr->qlock already taken by the thread. 
Furthermore,  **_osr_reap_done** tries to remove elements of osr->q, while **_txc_finish_io** iterates osr->q
So, I added **`if (txc->kv_submitted_sync && !txc->wal_txn)`** statement, for kv_sync_thread to call **_txc_finish calling _osr_reap_done**
<br/>
### w/ !wal_txn condition in _txc_state_proc on STATE_KV_DONE case:
Context1(ShardedOpWQ) -> **_txc_finish_io** -> _txc_finish_kv
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;pg lock&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**
Context2(kv_sync_thread) -> _kv_sync_thread-> **_txc_finish** calling **_osr_reap_done**
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**osr->qlock**
I think this way is okay, because kv_sync_thread isn't critical path in this case